### ISSUE TITLE
setItemPositions not working?

### ISSUE THREAD

**Issue by ssovit** [Arcana/node-dota2#480]
```
/* Other steam things were here..

*/
steamClient.on('logOnResponse', function (logonResp) {
        if (logonResp.eresult === Steam.EResult.OK) {
            console.log('Logged in!');
            steamFriends.setPersonaState(Steam.EPersonaState.Online);

            steamWebLogOn.webLogOn(function (sessionID, newCookie) {
                cookies = newCookie;
                getSteamAPIKey({
                    sessionID: sessionID,
                    webCookie: newCookie
                }, function (err, _APIKey) {
                    offers.setup({
                        sessionID: sessionID,
                        webCookie: newCookie,
                        APIKey: _APIKey
                    });
                    APIKey = _APIKey;
                    sortInventory();
                });
                setTimeout(function () {
                    addPendingFriends();
                    addPendingGroups();
                    acceptAllTradeOffers(false);
                }, 10000);
            });
        }
    });
/* other things are here.. */
var sortInventory = function () {
        var itemUrl = "http://api.steampowered.com/IEconItems_570/GetPlayerItems/v1/?key=" + APIKey + "&steamid=" + secrets.profileId;
        request(itemUrl, function (err, response, body) {
            if (err) {
                console.log("failed to retrive bot inventory")
                return false;
            }
            var resp = JSON.parse(body),
                items = [],
                position = [];
            if (resp.result.status == 1) {
                items = resp.result.items;
                var positions = sortItems(items);
                d2.launch();
                d2.on("ready", function () {
                    d2.setItemPositions(positions);
                    setTimeout(function () {
                        d2.exit();
                    }, 10000);
                });
                /* d2.on('inventoryUpdate', function () {
                     d2.setItemPositions(positions);

                 });*/
            }
            //return false;
        });
    };

    var sortItems = function (items) {
        var stack = [];
        var positions = [];
        var position = 0;
        //fs.writeFileSync(logFile + "-inventory-original.json", JSON.stringify(items));
        // group similar items together
        for (i = 0; i < items.length; i++) {

            if (typeof stack[items[i].defindex] == "undefined") {
                stack[items[i].defindex] = [];
            }
            stack[items[i].defindex].push(items[i].id);
        }
        stack.sort();
        _.each(stack, function (aitems, def) {
            _.each(aitems, function (itemid) {
                //positions.push([itemid, ++position]); // this one used to work before
                positions.push({ item_id: itemid, position: ++position }); // this one is new according to doc but not working
            });
        });
        //fs.writeFileSync(logFile + "-inventory-positions.json", JSON.stringify(positions));

        return positions;
    };
```

**Comment by ssovit** [comment]
hmm.. I made it working by adding these to my bot. Override setItemPositions made to ignore protobuff's illegal buffer error 

```
Dota2.Dota2Client.prototype.setItemPositions = function (item_positions) {
    console.log("Setting item positions.");
    let protoMask = 0x80000000;

    let payload = {
        "item_positions": item_positions
    };
    this.sendToGC((1077 | protoMask),
        Dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};


process.on('uncaughtException', function (exception) {
    // handle or ignore error
});
```

not sure if this is proper way to do.. but its working for me

**Comment by Crazy-Duck** [comment]
Could you verify if this works?
```
Dota2.Dota2Client.prototype.setItemPositions = function(item_positions) {
    if (this.debug) util.log("Setting item positions.");
    
    let payload = {
        "itemPositions": item_positions
    };
    this._gc.send(  {
                        "msg":Dota2.schema.lookupEnum("EDOTAGCMsg").values.k_EMsgGCSetItemPositions
                    }, 
                    Dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```

**Comment by ssovit** [comment]
@Crazy-Duck that didn't work.

But this one works
```
dota2.Dota2Client.prototype.setItemPositions = function (item_positions) {
    console.log("Setting item positions.");
    let protoMask = 0x80000000;

    let payload = {
        "item_positions": item_positions
    };
    this.sendToGC(Dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSetItemPositions,
        dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```

`k_EMsgGCSetItemPositions` is in `EGCItemMsg` but not in `EDOTAGCMsg` in [econ_gcmessages.proto](https://github.com/Arcana/node-dota2/blob/staging/proto/econ_gcmessages.proto#L63)

but that keeps throwing buffer error so I had to add 
```
process.on('uncaughtException', function (exception) {
    // handle or ignore error
});
```
to not crash the code as temporary solution.

**Comment by ssovit** [comment]
```
let payload = {
        "item_positions": item_positions
};
```
This one worked instead of `"itemPositions": item_positions` as in schema file [base_gcmessages.proto](https://github.com/Arcana/node-dota2/blob/staging/proto/base_gcmessages.proto#L302)

**Comment by Crazy-Duck** [comment]
Oh right, my bad, that one was still in the (very) old notation when protobufjs returned everything in camelCase instead of snake_case :blush: To be honest, since I've got no code to test this on my own and noone complained, I haven't looked into this for a long time and it's gotten a bit outdated... Could you try this?
```
Dota2.Dota2Client.prototype.setItemPositions = function(item_positions) {
    if (this.debug) util.log("Setting item positions.");
    
    let payload = {
        "item_positions": item_positions
    };
    this._gc.send(  {
                        "msg":Dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSetItemPositions
                    }, 
                    Dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```
I'm trying to avoid having to encode a protomask in this function

**Comment by ssovit** [comment]
not working

**Comment by Crazy-Duck** [comment]
Wait, so this (copied from your comment above) works?
```
dota2.Dota2Client.prototype.setItemPositions = function (item_positions) {
    console.log("Setting item positions.");
    let protoMask = 0x80000000;

    let payload = {
        "item_positions": item_positions
    };
    this.sendToGC(Dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSetItemPositions,
        dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```

**Comment by Crazy-Duck** [comment]
Because you're no longer using the protoMask in that one

**Comment by ssovit** [comment]
```
dota2.Dota2Client.prototype.setItemPositions = function (item_positions) {
    console.log("Setting item positions.");
    let protoMask = 0x80000000;

    let payload = {
        "item_positions": item_positions
    };
    this.sendToGC(Dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSetItemPositions,
        dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```

Oh sorry.. thought this was not working but there was no inventory sorting done. Sorry for confusion

**Comment by ssovit** [comment]
`    this.sendToGC((1077 | protoMask), dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());`
this one is only one working right now

**Comment by ssovit** [comment]
This is what I see when error is happening https://screencast.ssovit.com/2017-04-18_15h41_40.png

**Comment by Crazy-Duck** [comment]
Wait, that error is produced in an entirely different part of the code! The error you're getting is produced in handlers\cache.js upon receiving the new snapshot of the inventory D:

**Comment by ssovit** [comment]
ya.. when its not working no errors are shown, but when its working it throws error

**Comment by Crazy-Duck** [comment]
Can you just run the library without doing your inventory stuff? There seems to be something particular about your inventory that crashes the parsing of the cache update. When I run example I get the following output:
```
18 Apr 10:18:30 - Logged on.
18 Apr 10:18:30 - Launching Dota 2
18 Apr 10:18:30 - Received servers.
18 Apr 10:18:31 - Sending ClientHello
18 Apr 10:18:32 - Received client welcome.
18 Apr 10:18:32 - Node-dota2 ready.
18 Apr 10:18:32 - Cache(s) subscribed, type(s): 1,7
18 Apr 10:18:32 - Received inventory snapshot
18 Apr 10:18:32 - Unhandled cache ID: 7
```
So you see that my inventory is parsed without issues. Could you try and run the lib without doing anything as well and check if the error shows up?

**Comment by ssovit** [comment]
ya tried that.. if I don't call `setItemPositions` no error shows up.

**Comment by Crazy-Duck** [comment]
hmm, so it's the inventory update after calling setItemPositions that causes issues?

**Comment by ssovit** [comment]
I am not sure ;)

**Comment by Crazy-Duck** [comment]
should be easy enough to figure out, just put your setItemPositions in a time-out :p

**Comment by ssovit** [comment]
yeah.. i will do some checks.. i will keep you updated if I find a working solution

**Comment by ssovit** [comment]
This one worked
```
dota2.Dota2Client.prototype.setItemPositions = function (item_positions) {
    console.log("Setting item positions.");
    let payload = {
        "item_positions": item_positions
    };
    var cache = dota2.Dota2Client.prototype._handlers[dota2.schema.lookupEnum("ESOMsg").values.k_ESOMsg_UpdateMultiple];
    dota2.Dota2Client.prototype._handlers[dota2.schema.lookupEnum("ESOMsg").values.k_ESOMsg_UpdateMultiple] = function () {
        dota2.Dota2Client.prototype._handlers[dota2.schema.lookupEnum("ESOMsg").values.k_ESOMsg_UpdateMultiple] = cache;
    };
    this.sendToGC(dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSetItemPositions, dota2.schema.lookupType("CMsgSetItemPositions").encode(payload).finish());
};
```
turned off `k_ESOMsg_UpdateMultiple` for once as data was not being parsed properly or data was incomplete as it was new items position value(I suppose) message from GC and restored immediately after

**Comment by Ne3tCode** [comment]
@Crazy-Duck 
In `onUpdateMultiple`: `obj.object_data` is a `Buffer`, but `handleSubscribedType` expects `Buffer[]` (array).
As this code may be applied for a single item, you cannot use `this.Inventory = inv;` in `handleSubscribedType`.

**Comment by Crazy-Duck** [comment]
Alright, so if understand correctly, inventory caches are handled differently than Lobby/Chat caches, who get sent in their entirety each time there's an update?

**Comment by ssovit** [comment]
Its as @Ne3tCode said
on first "inventoryUpdate", it gives `[<Buffer ..... >, <Buffer ... >,...]` 
but when setItemPositions if fired, it gives just `<Buffer ... >`, just the item that was set to new position(fires handleSubscribedType for all items individually).

handlers/cache.js
```
function handleSubscribedType(obj_type, object_data) {
    switch (obj_type) {
        // Inventory item
    case cacheTypeIDs.CSOEconItem:
        if (this.debug) util.log("Received inventory snapshot");
        var doEmit = true;
        if (object_data instanceof Buffer) {
            object_data = [object_data];
            doEmit = false
        }
        var inv = object_data.map(obj => Dota2.schema.lookupType("CSOEconItem").decode(obj));
        if (doEmit) {
            this.emit("inventoryUpdate", inv);
            this.Inventory = inv;
        }
        break;
.....```
this patch did prevent triggering inventoryUpdate when object_data was just instance of Buffer when setItemPositions is fired.

with some digging another call called `sortItems` like d2.sortItems(2)
```
dota2.Dota2Client.prototype.sortItems = function (type) {
    console.log("sorting Items");
    let payload = {
        "sort_type": type
    };
   
    this.sendToGC(dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSortItems, dota2.schema.lookupType("CMsgSortItems").encode(payload).finish());
};
```
for me right now `type=2` worked best for my case which would sort items and keep hero set items together. Not sure what could be possible sortItems parameters.
This also has gives same error as setItemPositions so instance check of object_data did the fix for me.

**Comment by Crazy-Duck** [comment]
Alright, I'm writing a fix for that as we speak, but the problem's gonna be a little broader, since I'll probably need to take into account updateMultiples where there's additions and removals

**Comment by Crazy-Duck** [comment]
I pushed a new branch called items, can you check with that one if everything's working?

**Comment by ssovit** [comment]
https://screencast.ssovit.com/2017-04-19_18h17_58.png
got this error

**Comment by ssovit** [comment]
Here is my bot https://github.com/ssovit/Dota2Dispenser

**Comment by Crazy-Duck** [comment]
My bad, forgot two brackets, think it should be better now :crossed_fingers:

**Comment by ssovit** [comment]
Perfect!!

**Comment by ssovit** [comment]
BTW.

```
dota2.Dota2Client.prototype.sortItems = function (type) {
    console.log("sorting Items");
    let payload = {
        "sort_type": type
    };
    this.sendToGC(dota2.schema.lookupEnum("EGCItemMsg").values.k_EMsgGCSortItems, dota2.schema.lookupType("CMsgSortItems").encode(payload).finish());
};
```
This would be good addition to inventory.js as its fast and won't need sorting array and making it more faster. But not sure about what could be possible values for `type`

like sorting items by rarity, heroes, and so on like in dota 2 client.

**Comment by Crazy-Duck** [comment]
Can you create a new issue for that one? That way I can close this issue with the PR into staging. I'll check sortItems and see if I can find out a bit more about it

**Comment by ssovit** [comment]
Sure.. Thanks for your hard work 🥇

**Comment by markrmessmore** [comment]
I came across this post as I'm experiencing the same error with the **Illegal buffer**. Given I've been coding a whopping three months I can tell you guys resolved the issue however I'm having trouble following how the resolution was implemented. 

I see @Crazy-Duck mentioned a branch called 'items' but that seems to no longer exist. I'm using version 5.1.0 of node-dota2 so I guess my first question is...is this the same error showing up for a different reason? Has a 'full fix' not been implemented? How do I resolve my own Illegal Buffer? Why is monosyllable such a long word? That sort of stuff.

Any help/guidance would be appreciated.
